sbi/diagnostics/tarp.py

```diff
@@ -0,0 +1,324 @@
+        samples = samples.detach()
+
+        num_samples = samples.shape[0]  # samples per simulation
+        num_sims = samples.shape[-2]
+        num_dims = samples.shape[-1]
+
+        if self.n_bins is None:
+            self.n_bins = num_sims // 10
+
+        if theta.shape[-2] != num_sims:
+            raise ValueError("theta must have the same number of rows as samples")
+        if theta.shape[-1] != num_dims:
+            raise ValueError("theta must have the same number of columns as samples")
+
+        if self.do_norm:
+            lo = torch.min(theta, dim=-2, keepdims=True).values  # min along num_sims
+            hi = torch.max(theta, dim=-2, keepdims=True).values  # max along num_sims
+            samples = (samples - lo) / (hi - lo + 1e-10)
+            theta = (theta - lo) / (hi - lo + 1e-10)
+
+        assert len(theta.shape) == len(samples.shape)
+
+        if not isinstance(self.references, Tensor):
+            # obtain min/max per dimension of theta
+            lo = (
+                torch.min(theta, dim=-2).values.min(axis=0).values
+            )  # should be 0 if normalized
+            hi = (
+                torch.max(theta, dim=-2).values.max(axis=0).values
+            )  # should be 1 if normalized
+
+            refpdf = torch.distributions.Uniform(low=lo, high=hi)
+            self.references = refpdf.sample((1, num_sims))
+        else:
+            if len(self.references.shape) == 2:
+                # add singleton dimension in front
+                self.references = self.references.unsqueeze(0)
+
+            if len(self.references.shape) == 3 and self.references.shape[0] != 1:
+                raise ValueError(
+                    f"""references must be a 2D array with a singular first
+                    dimension, received {self.references.shape}"""
+                )
+
+            if self.references.shape[-2] != num_sims:
+                raise ValueError(
+                    f"references must have the same number samples as samples,"
+                    f"received {self.references.shape[-2]} != {num_sims}"
+                )
+
+            if self.references.shape[-1] != num_dims:
+                raise ValueError(
+                    "references must have the same number of dimensions as "
+                    f"samples or theta, received {self.references.shape[-1]}"
+                    f"!= {num_dims}"
+                )
+
+        assert len(self.references.shape) == len(
+            samples.shape
+        ), f"references {self.references.shape} != samples {samples.shape}"
+
+        if self.metric_name.lower() in ["l2", "euclidean"]:
+            distance = l2
+        elif self.metric_name.lower() in ["l1", "manhattan"]:
+            distance = l1
+        else:
+            raise ValueError(
+                "metric must be either 'euclidean' or 'manhattan',"
+                f"received {self.metric_name}"
+            )
```

Comment:
this could be done during `__init__` as well, and then just set `self.distance`.

Reply:
we decided to discard the implementation as a class. So this is not required anymore.